### PR TITLE
rgw: Add a wrapper for librados::AioCompletion to prevent memory leaks

### PR DESCRIPTION
**Pull request by vedanshbhartia**
librados::AioCompletion is created at multiple places in rgw code, but the developer might forget to delete the completion in case of errors. This PR introduces a wrapper over the completion. The wrapper is supposed to be used as a unique pointer, which will cause the completion to be destroyed when it goes out of scope.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by yuvalif** [comment]
jenkins test api

File: src/rgw/driver/rados/rgw_gc.cc [hunk 1]
```diff
@@ -170,9 +170,8 @@ void RGWGC::on_defer_canceled(const cls_rgw_gc_obj_info& info)
   cls_rgw_gc_queue_defer_entry(op, cct->_conf->rgw_gc_obj_min_wait, info);
   cls_rgw_gc_remove(op, {tag});
 
-  auto c = librados::Rados::aio_create_completion(nullptr, nullptr);
-  store->gc_aio_operate(obj_names[i], c, &op);
-  c->release();
+  auto c = make_unique<RgwAioCompletionWrapper>(nullptr, nullptr);
```
Comment (yuvalif) [inline src/rgw/driver/rados/rgw_gc.cc:173]: why do we need a `unique_ptr` here?
why not just use the completion wrapper?

File: src/rgw/rgw_common.h [hunk 2]
```diff
@@ -1839,3 +1839,32 @@ rgw_global_init(const std::map<std::string,std::string> *defaults,
 		    std::vector < const char* >& args,
 		    uint32_t module_type, code_environment_t code_env,
 		    int flags);
+
+class RgwAioCompletionWrapper {
```
Comment (cbodley) [inline src/rgw/rgw_common.h:1843]: i really like the use of [RAII](https://en.cppreference.com/w/cpp/language/raii) to handle this cleanup 👍

but `RgwAioCompletionWrapper` is essentially just reinventing [std::unique_ptr](https://en.cppreference.com/w/cpp/memory/unique_ptr). if we add a custom deleter for `AioCompletion`, we can use `std::unique_ptr` directly:
```c++
// unique_ptr deleter that calls release()
struct AioCompletionDeleter {
  void operator()(librados::AioCompletion* c) { c->release(); }
};
using aio_completion_ptr = std::unique_ptr<librados::AioCompletion, AioCompletionDeleter>;
```
callers could either construct this ptr directly:
```c++
  aio_completion_ptr c{librados::Rados::aio_create_completion(nullptr, nullptr)};
```
or we could add a helper function that returns `aio_completion_ptr`:
```c++
  auto c = make_aio_completion(nullptr, nullptr);
```
what do you think?

**Review by yuvalif — APPROVED** [review]

**Comment by yuvalif** [comment]
unstale please

File: src/rgw/rgw_common.h (hunk 2 above)
Comment (yuvalif) [reply src/rgw/rgw_common.h]: @cbodley could you please re-review?

**Review by cbodley — APPROVED** [review]

**Comment by yuvalif** [comment]
teuthology results: https://pulpito.ceph.com/yuvalif-2023-12-17_12:10:21-rgw-wip-yuval-test-gsoc-distro-default-smithi/
failures:
* lua - expected
* multisite - expected
* `install python-dev'` - expected
* this fialure:
```
raceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_5f3731ee8689ef3b0ef7d571e4cc2cb968c78d73/teuthology/run_tasks.py", line 105, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_teuthology_5f3731ee8689ef3b0ef7d571e4cc2cb968c78d73/teuthology/run_tasks.py", line 83, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_61a02478093e1f8a9a94ec9939d540b293d6efc3/qa/tasks/radosgw_admin_rest.py", line 230, in task
    assert out['truncated'] == False
AssertionError
```
is also seen on other branches

**Pull request merged**